### ISSUE TITLE
Order being flagged for review by Radar causes double stock reduction

### ISSUE THREAD

**Issue by mikkamp** [woocommerce/woocommerce-gateway-stripe#981]
#### Affected ticket(s)
2319749-zen

#### What I expected
When an order is flagged for review by a Radar rule, the order process should be changed to on-hold, until the review has been cleared. Even though the order passes through two status changes I'd expect the stock to be reduced only once.

#### What happened instead
The order status is initially changed to processing which triggers a possible stock reduction. We also receive a second request for a review_opened webhook. This triggers the order to change to on-hold, which also triggers a possible stock reduction.

WooCommerce should prevent a double stock reduction on the same order. However since there is a race condition between the two requests, the order object isn't saved yet before the next one checks if the stock needs to be reduced.

#### Steps to reproduce the issue
I'm not able to reproduce a double stock reduction on my test site (because the timing is slightly different in the race condition). However I can reproduce odd behaviour where the order notes indicate that the status changed to on-hold but the status remains processing.

1\. Create a radar rule for placing a payment in review, I used the following:
Review if `:amount_in_usd: > 100.00`
2\. Test an order which would be flagged, total amount in order over $100
3\. Observe the order notes
In my case the following order note is shown:
> A review has been opened for this order. Action is needed. Please go to your Stripe Dashboard to review the issue. Reason: (rule) Order status changed from Processing to On hold.

But the order remains in processing. In this case it didn't reproduce the double stock reduction, but it does demonstrate there are two requests handling the order object.

I added a `sleep(2)` when processing the review_opened webhook and then repeated the same test. This ensured the two requests didn't happen at the same time, and allowed the order status to correctly change to on-hold.

Here is some logging which was provided which shows how the double stock reduction occurs:

Processing order request:

```
2019-09-02 00:12:59 148114 (148103) ESFWSWI15GRYBLK-8-uk 0
Plugin: woocommerce wc_update_product_stock() via wc_update_product_stock() from wc_reduce_stock_levels()
[BT]
1. WP_Hook->apply_filters(,Array) in /wp-includes/class-wp-hook.php:310
2. WP_Hook->do_action(Array) in /wp-includes/plugin.php:465
3. do_action(woocommerce_variation_set_stock,WC_Product_Variation object) in /woocommerce/includes/wc-stock-functions.php:55
4. wc_update_product_stock(WC_Product_Variation object,1,decrease) in /woocommerce/includes/wc-stock-functions.php:175
5. wc_reduce_stock_levels(171628) in /woocommerce/includes/wc-stock-functions.php:100
6. wc_maybe_reduce_stock_levels(171628) in /wp-includes/class-wp-hook.php:288
7. WP_Hook->apply_filters(Array) in /wp-includes/class-wp-hook.php:310
8. WP_Hook->do_action(Array) in /wp-includes/plugin.php:465
9. do_action(woocommerce_order_status_processing,171628,WC_Order object) in /woocommerce/includes/class-wc-order.php:360
10. WC_Order->status_transition() in /woocommerce/includes/class-wc-order.php:230
11. WC_Order->save() in /woocommerce/includes/class-wc-order.php:119
12. WC_Order->payment_complete(ch_1FE3ACBqpVHQgfQ797gqVceb) in /woocommerce-gateway-stripe/includes/abstracts/abstract-wc-stripe-payment-gateway.php:436
13. WC_Stripe_Payment_Gateway->process_response(stdClass object,WC_Order object) in /woocommerce-gateway-stripe/includes/class-wc-gateway-stripe.php:670
14. WC_Gateway_Stripe->process_payment(171628) in /woocommerce/includes/class-wc-checkout.php:923
15. WC_Checkout->process_order_payment(171628,stripe) in /woocommerce/includes/class-wc-checkout.php:1123
16. WC_Checkout->process_checkout() in /woocommerce/includes/class-wc-ajax.php:462
17. WC_AJAX::checkout() in /wp-includes/class-wp-hook.php:286
18. WP_Hook->apply_filters(,Array) in /wp-includes/class-wp-hook.php:310
19. WP_Hook->do_action(Array) in /wp-includes/plugin.php:465
20. do_action(wc_ajax_checkout) in /woocommerce/includes/class-wc-ajax.php:87
21. WC_AJAX::do_wc_ajax() in /wp-includes/class-wp-hook.php:286
22. WP_Hook->apply_filters(,Array) in /wp-includes/class-wp-hook.php:310
23. WP_Hook->do_action(Array) in /wp-includes/plugin.php:465
24. do_action(template_redirect) in /wp-includes/template-loader.php:13
25. require_once(/wp-includes/template-loader.php) in /wp-blog-header.php:19
26. require(/wp-blog-header.php) in /home/super8/public_html/index.php:17
```

And the second request for the review_opened webhook which happens at the same time:

```
2019-09-02 00:12:59 148114 (148103) ESFWSWI15GRYBLK-8-uk -1
Plugin: woocommerce wc_update_product_stock() via wc_update_product_stock() from wc_reduce_stock_levels()
[BT]
1. WP_Hook->apply_filters(,Array) in /wp-includes/class-wp-hook.php:310
2. WP_Hook->do_action(Array) in /wp-includes/plugin.php:465
3. do_action(woocommerce_variation_set_stock,WC_Product_Variation object) in /woocommerce/includes/wc-stock-functions.php:55
4. wc_update_product_stock(WC_Product_Variation object,1,decrease) in /woocommerce/includes/wc-stock-functions.php:175
5. wc_reduce_stock_levels(171628) in /woocommerce/includes/wc-stock-functions.php:100
6. wc_maybe_reduce_stock_levels(171628) in /wp-includes/class-wp-hook.php:288
7. WP_Hook->apply_filters(Array) in /wp-includes/class-wp-hook.php:310
8. WP_Hook->do_action(Array) in /wp-includes/plugin.php:465
9. do_action(woocommerce_order_status_on-hold,171628,WC_Order object) in /woocommerce/includes/class-wc-order.php:360
10. WC_Order->status_transition() in /woocommerce/includes/class-wc-order.php:230
11. WC_Order->save() in /woocommerce/includes/class-wc-order.php:334
12. WC_Order->update_status(on-hold,A review has been opened for this order. Action is needed. Please go to your Stripe Dashboard to review the issue. Reason: (rule)) in /woocommerce-gateway-stripe/includes/class-wc-stripe-webhook-handler.php:559
13. WC_Stripe_Webhook_Handler->process_review_opened(stdClass object) in /woocommerce-gateway-stripe/includes/class-wc-stripe-webhook-handler.php:736
14. WC_Stripe_Webhook_Handler->process_webhook({...}) in /woocommerce-gateway-stripe/includes/class-wc-stripe-webhook-handler.php:69
15. WC_Stripe_Webhook_Handler->check_for_webhook() in /wp-includes/class-wp-hook.php:286
16. WP_Hook->apply_filters(,Array) in /wp-includes/class-wp-hook.php:310
17. WP_Hook->do_action(Array) in /wp-includes/plugin.php:465
18. do_action(woocommerce_api_wc_stripe) in /woocommerce/includes/class-wc-api.php:113
19. WC_API->handle_api_requests(WP object) in /wp-includes/class-wp-hook.php:286
20. WP_Hook->apply_filters(Array) in /wp-includes/class-w
```

A temporary workaround is to use the following filter to prevent the order from changing to on-hold:

```
add_filter( 'wc_stripe_webhook_review_change_order_status', '__return_false' );
```

**Comment by tommyshellberg** [comment]
Just updating here that it's still an issue with 4.3 Alpha:

![https://cld.wthms.co/ODVzy8](https://cld.wthms.co/ODVzy8+)
Image link: https://cld.wthms.co/ODVzy8

**Comment by skullo27** [comment]
We are having the exact same issue. In our case, we cannot reproduce the issue on our staging server which suggests agreement with the above comments on timing. Our issue is for a Radar rule that places the order in a review if the address zip check fails. We have a number of conditions active in stripe's radar, but this is the only one where the issue occurs. Furthermore, it is not consistent and does not happen on every order that is placed on hold but the rule, however, it is the only common factor that we can observe from orders exhibiting this issue.

**Comment by v18** [comment]
I am not able to repro the incorrect order status. The order status is correctly `on-hold` for me with these steps:

1. Create a radar rule for placing a payment in review, I used the following:
Review if :amount_in_usd: > 100.00
2. Test an order which would be flagged, total amount in order over $100
3. Observe the order notes
In my case the following order note is shown:
> A review has been opened for this order. Action is needed. Please go to your Stripe Dashboard to review the issue. Reason: (rule) Order status changed from Processing to On hold.

However, the order is correctly set to status `on-hold` at this point.

@skullo27 are you experiencing this issue still?

I'm going to set it to priority low for now, and update as needed when we hear back.

**Comment by mikkamp** [comment]
I no longer have the same environment setup to retest it. But I wouldn't be surprised if it works differently after these changes: https://github.com/woocommerce/woocommerce/blob/5.0.0/changelog.txt#L722

Maybe a comparison test of WC 4.2 and 5.0 would help. The WC version that matches the timing of the last comment would be 4.1.

**Comment by thenbrent** [comment]
Closing this one as there has been no activity in >2.5 years and the above comment suggests that may be explained by changes to WC core addressing the issue.
